pennylane/transforms/transpile.py

```diff
@@ -0,0 +1,156 @@
+
+
+@qfunc_transform
+def transpile(tape: QuantumTape, coupling_map: Union[List, nx.Graph]):
+    """ Transpile a circuit according to a desired coupling map
+
+    Args:
+        tape (function): A quantum function.
+        coupling_map (list[tuple(int, int)] or nx.Graph): Either a list of tuples(int, int) or an instance of
+            networkx.Graph specifying the couplings between different qubits.
+    Returns:
+        function: the transformed quantum function
+
+    **Example**
+
+    Consider the following example circuit
+
+    .. code-block:: python
+
+        def circuit():
+            qml.CNOT(wires=[0, 1])
+            qml.CNOT(wires=[2, 3])
+            qml.CNOT(wires=[1, 3])
+            qml.CNOT(wires=[1, 2])
+            qml.CNOT(wires=[2, 3])
+            qml.CNOT(wires=[0, 3])
+            return qml.probs(wires=[0, 1, 2, 3])
+
+    which, before transpiling it looks like this:
+
+        0: ──╭C──────────────╭C──╭┤ Probs
+        1: ──╰X──╭C──╭C──────│───├┤ Probs
+        2: ──╭C──│───╰X──╭C──│───├┤ Probs
+        3: ──╰X──╰X──────╰X──╰X──╰┤ Probs
+
+    Suppose we have a device which has connectivity constraints according to the graph:
+
+        0 --- 1
+        |     |
+        2 --- 3
+
+    We encode this in a coupling map as a list of the edges which are present in the graph, and then pass this, together
+    with the circuit, to the transpile function to get a circuit which can be executed for the specified coupling map
+
+    >>> dev = qml.device('default.qubit', wires=[0, 1, 2, 3])
+    >>> transpiled_circuit = qml.transforms.transpile(coupling_map=[(0, 1), (1, 3), (3, 2), (2, 0)])(circuit)
+    >>> transpiled_qnode = qml.QNode(circuit, dev)
+    >>> print(qml.draw(transpiled_qnode))
+
+     0: ──╭C─────────────────────╭C──╭┤ Probs
+     1: ──╰X──╭C─────────╭C──────│───├┤ Probs
+     2: ──╭C──│───╭SWAP──│───╭X──╰X──├┤ Probs
+     3: ──╰X──╰X──╰SWAP──╰X──╰C──────╰┤ Probs
+
+    A swap gate has been applied to wires 2 and 3, and the remaining gates have been adapted accordingly
+
+    """
+    # init connectivity graph
+    coupling_graph = nx.Graph(coupling_map) if not isinstance(coupling_map, nx.Graph) else coupling_map
+
+    # make sure every wire is present in coupling map
+    if any([wire not in coupling_graph.nodes for wire in tape.wires]):
+        raise ValueError(
+            f'not all wires present in coupling map! wires: {tape.wires}, coupling map: {coupling_graph.nodes}')
+
+    def _adjust_op_indices(_op, _map_wires):
+        """ helper function which adjusts wires in Operation according to the map _map_wires"""
+        _new_wires = Wires([_map_wires[w] for w in _op.wires])
+        _params = _op.parameters
+        if len(_params) == 0:
+            return type(_op)(wires=_new_wires)
+        return type(_op)(*_params, wires=_new_wires)
+
+    def _adjust_mmt_indices(_m, _map_wires):
+        """ helper function which adjusts wires in MeasurementProcess according to the map _map_wires"""
```

Comment:
It sounds like this functionality is a useful utility we should add to PL more explicitly! Maybe something like `new_op = qml.map_wires(op, mapping)` (out of scope for this PR though :) )